app/gui/view/graph-editor/src/free_place_finder.rs

```diff
@@ -0,0 +1,255 @@
+
+/// The structure describing an occupied area.
+///
+/// All such areas are rectangles described by x and y ranges.
+#[allow(missing_docs)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct OccupiedArea {
+    pub x1: f32,
+    pub y1: f32,
+    pub x2: f32,
+    pub y2: f32,
+}
+
+impl OccupiedArea {
+    /// Get x position of the left boundary.
+    pub fn left(&self) -> f32 {
+        min(self.x1, self.x2)
+    }
+
+    /// Get x position of the right boundary.
+    pub fn right(&self) -> f32 {
+        max(self.x1, self.x2)
+    }
+
+    /// Get y position of the top boundary.
+    pub fn top(&self) -> f32 {
+        max(self.y1, self.y2)
+    }
+
+    /// Get y position of the bottom boundary.
+    pub fn bottom(&self) -> f32 {
+        min(self.y1, self.y2)
+    }
+
+    /// Check if the rectangle contains given point.
+    ///
+    /// The boundaries are open - they are not considered occupied.
+    pub fn contains(&self, point: Vector2) -> bool {
+        (self.x1 - point.x) * (self.x2 - point.x) < 0.0
+            && (self.y1 - point.y) * (self.y2 - point.y) < 0.0
+    }
+
+    /// Return the x position of the left or right boundary, depending on whether the direction
+    /// points leftwards of rightwards respectively. Returns `None` if direction does not point
+    /// leftwards nor rightwards.
+    pub fn x_bound_following_direction(&self, direction: Vector2) -> Option<f32> {
+        if direction.x > f32::EPSILON {
+            Some(self.right())
+        } else if direction.x < -f32::EPSILON {
+            Some(self.left())
+        } else {
+            None
+        }
+    }
+
+    /// Return the y position of the top or bottom boundary, depending on whether the direction
+    /// points toward the top or bottom respectively. Returns `None` if direction does not point
+    /// toward top nor bottom.
+    pub fn y_bound_following_direction(&self, direction: Vector2) -> Option<f32> {
+        if direction.y > f32::EPSILON {
+            Some(self.top())
+        } else if direction.y < -f32::EPSILON {
+            Some(self.bottom())
+        } else {
+            None
+        }
+    }
+
+    /// Return the point where the ray going from starting point along the `direction` vector will
+    /// intersect with area boundaries, if there is exactly one such point, otherwise the result is
+    /// unspecified.
+    pub fn boundary_intersection(&self, starting_point: Vector2, direction: Vector2) -> Vector2 {
+        let x_bound = self.x_bound_following_direction(direction);
+        let dir_x_factor = x_bound.map(|x| (x - starting_point.x) / direction.x);
+        let y_bound = self.y_bound_following_direction(direction);
+        let dir_y_factor = y_bound.map(|y| (y - starting_point.y) / direction.y);
+        let dir_factor = match (dir_x_factor, dir_y_factor) {
+            (Some(x), Some(y)) => min(x, y),
+            (Some(x), None) => x,
+            (None, Some(y)) => y,
+            _ => default(),
+        };
```

Comment:
I'm wondering if we can have any nicer way of expressing such things, I see this code in different places of the codebase. In Haskell, these 6 lines would be 
```
unwrap_or_default (min <$> dir_x_factor <*> dir_y_factor)
```

I'm wondering if we should not introduce Applicative Functors too (basically these 2 operators above).

Reply:
TBH, the six-line version seems more readable. I'm not a Haskell person, however.